### ISSUE TITLE
HttpClientFactory creates a HttpClient that is not compatible with ProgressMessageHandler

### ISSUE THREAD

**Issue by StephenRieger** [xamarin/xamarin-macios#5636]
### Steps to Reproduce

1. Configure the project to use NSURLSessionHandler as the default handler.
2. Add the Microsoft.AspNet.WebApi.Client package.
3. Create a progress message handler:

            ProgressMessageHandler progressMsgHander = new ProgressMessageHandler();
            progressMsgHander.HttpSendProgress += (object sender, HttpProgressEventArgs e) =>
            {
                ...
            };
4. Create an HttpClient that uses that handler:

            HttpClient = HttpClientFactory.Create(progressMsgHander);

5. Use that client to upload data from a FileStream:
                
                using (StreamContent streamContent = new StreamContent(fileStream))
                {
                    streamContent.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
                    streamContent.Headers.ContentLength = fileStream.Length;

                    HttpResponseMessage response = await HttpClient.PostAsync(address, streamContent);
                    response.EnsureSuccessStatusCode();

                    return new Response { ResponseText = await response.Content.ReadAsStringAsync() 
                };

### Expected Behavior
The HttpSendProgress handler is progressively called as the file is uploaded
### Actual Behavior
The HttpSendProgress handler is called before the actual upload, moving from 0 to 100%, followed by the actual upload.

The root cause looks like this bit in NSUrlRequest.CreateRequest:

if (request.Content != null) {
                    stream = await request.Content.ReadAsStreamAsync ().ConfigureAwait (false);
                    headers = headers.Union (request.Content.Headers).ToArray ();
}

where request.Content.ReadAsStreamAsync makes a copy of the fileStream that was passed to the StreamContent constructor.

It looks to me like this could be fixed by calling CreateContentReadStreamAsync() instead of ReadAsStreamAsync() as StreamContent overrides that to return a read-only copy of the content stream. (https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/StreamContent.cs)

This likely also breaks the attempt to avoid buffering large files in memory (MaxInputInMemory and WrappedNSInputStream) as the file will have already been copied to an in memory stream by the time that code is reached.
### Environment

```
=== Visual Studio Community 2017 for Mac ===

Version 7.7.4 (build 1)
Installation UUID: e561e3bb-b8c1-415a-8d73-99a4cb65fe09
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 4.4.1.178 (master / eeaeb7e6)

	Package version: 516000221

=== Mono Framework MDK ===

Runtime:
	Mono 5.16.0.221 (2018-06/b63e5378e38) (64-bit)
	Package version: 516000221

=== NuGet ===

Version: 4.8.0.5385

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	2.1.8
	2.1.7
	2.1.4
	2.1.2
	2.0.5
SDK: /usr/local/share/dotnet/sdk/2.1.504/Sdks
SDK Versions:
	2.1.504
	2.1.503
	2.1.402
	2.1.302
	2.1.4
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/5.16.0/lib/mono/msbuild/15.0/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.4
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Xamarin.Android ===

Version: 9.1.8.0 (Visual Studio Community)
Android SDK: /Users/srieger/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 28.0.0
SDK Build Tools Version: 27.0.3

=== Microsoft Mobile OpenJDK ===

Java SDK: /Users/srieger/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_1.8.0.9
openjdk version "1.8.0-9"
OpenJDK Runtime Environment (build 1.8.0-9-microsoft-b00)
OpenJDK 64-Bit Server VM (build 25.71-b00, mixed mode)

Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android Device Manager ===

Version: 7.7.5.0
Hash: 06ceaea1

=== Apple Developer Tools ===

Xcode 10.1 (14460.46)
Build 10B61

=== Xamarin Inspector ===

Version: 1.4.3
Hash: db27525
Branch: 1.4-release
Build date: Mon, 09 Jul 2018 21:20:18 GMT
Client compatibility: 1

=== Xamarin.Mac ===

Version: 5.2.1.15 (Visual Studio Community)
Hash: d60abd198
Branch: 
Build date: 2019-02-01 12:23:30-0500

=== Xamarin.iOS ===

Version: 12.2.1.15 (Visual Studio Community)
Hash: d60abd198
Branch: d15-9
Build date: 2019-02-01 12:23:29-0500

=== Build Information ===

Release ID: 707040001
Git revision: 2a4679454f31c8bb887cfd181ea133c9149d3cad
Build date: 2019-01-30 16:18:21+00
Build branch: release-7.7.4
Xamarin extensions: c8b4305fd4c60b9c25947072eb150fdc3a5aa2b5

=== Operating System ===

Mac OS X 10.14.3
Darwin 18.2.0 Darwin Kernel Version 18.2.0
    Thu Dec 20 20:46:53 PST 2018
    root:xnu-4903.241.1~1/RELEASE_X86_64 x86_64

```

**Comment by mandel-macaque** [comment]
@StephenRieger  What a beautiful issue report! Thanks a lot for the info. I will be taking a look at reproducing this and creating the required fix.

**Comment by mandel-macaque** [comment]
@StephenRieger I'm going to update the bug report title since I have confirmed that the behaviour you experience is the same with all the different handlers. Will keep investigating.

**Comment by mandel-macaque** [comment]
I have  used the following sample, which uses a memory stream large enough and I only see a single event in all the different MessageHandlers:

```csharp

public async override void ViewDidLoad ()
{
	base.ViewDidLoad ();
	ProgressMessageHandler progressMsgHander = new ProgressMessageHandler();
	progressMsgHander.HttpSendProgress += (object sender, HttpProgressEventArgs e) =>
	{
		Console.WriteLine ($"Sent {e.BytesTransferred}");
	};
	progressMsgHander.InnerHandler = new HttpClientHandler (); // you can change this to use the CFNetworkHandler and the NSUrlSessionHandler
	// Perform any additional setup after loading the view, typically from a nib.
	var httpClient = new HttpClient (progressMsgHander);
	var memoryStream = new MemoryStream (new byte [100000000]);
	using (StreamContent streamContent = new StreamContent(memoryStream))
	{
		streamContent.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
		streamContent.Headers.ContentLength = memoryStream.Length;

		HttpResponseMessage response = await httpClient.PostAsync("https://httpbin.org/post", streamContent);
		response.EnsureSuccessStatusCode();

		Console.WriteLine (await response.Content.ReadAsStringAsync ()); 
	};
}
```

**Comment by StephenRieger** [comment]
The progress handler is called once when using a MemoryStream, but is called multiple times when using a FileStream referencing a file on disk. Though with a large enough file, I get an exception thrown - either a request entity too large (when using NSUrlSessionHandler) or an ObjectDisposedException (HttpClientHandler). A 77MB file works, a 148MB file throws. With the one that throws, I'm seeing the sent messages logged for the entire file - likely because those are all coming from the call to request.Content.ReadAsStreamAsync(), before the actual send.

**Comment by mandel-macaque** [comment]
We need to revisit this after the nerworking changes in mono @baulig can you give us a comment?

**Comment by baulig** [comment]
I spent some time investigating this and it's the `HttpContent` that does the actual writing - regardless of the underlying `HttpClientHandler`.

`HttpContent.SerializeToStreamAsync()` calls `Stream.CopyToAsync()` via `StreamToStreamCopy.CopyAsync()` with a buffer size of 81920.

Since your source content is a `MemoryStream`, the implementation of `MemoryStream.CopyToAsync()` writes the entire thing using the already-allocated internal buffer.  The `ProgressMessageHandler.WriteAsync()` method is then called with the full internal buffer, so you only get an update message once.

I tested this with both the new `SocketsHttpHandler` as well as `NSUrlSessionHandler` and in both cases, the `HttpSendProgress` callback is called once.
